src/main/scala/karazin/scala/users/group/week4/homework/Homework.scala

```diff
@@ -67,24 +72,36 @@ object Homework:
       else if x > elem  then right contains x
       else              true
 
-    // Optional task
-    infix def remove(x: Int): IntSet = ???
+    infix def remove(x: Int): IntSet =
+      if (x < elem) NonEmpty(elem, left.remove(x), right)
+      else if (x > elem) NonEmpty(elem, left, right.remove(x))
+      else left ∪ right
 
     @targetName("union")
-    infix def ∪(that: IntSet): IntSet = ???
+    infix def ∪(that: IntSet): IntSet =
+      left ∪ right ∪ that.include(elem)
 
     @targetName("intersection")
-    infix def ∩(that: IntSet): IntSet = ???
+    infix def ∩(that: IntSet): IntSet =
+      if (that.contains(elem)) left ∩ right ∩ that
+      else left ∩ right ∩ that
 
     @targetName("complement")
-    infix def ∖(that: IntSet): IntSet = ???
+    infix def \(that: IntSet): IntSet =
+      if (that.contains(elem)) left \ right \ that
+      else NonEmpty(elem, left \ that, right \ that)
 
     @targetName("disjunctive union")
-    infix def ∆(that: IntSet): IntSet = ???
+    infix def ∆(that: IntSet): IntSet =
+      (left \ that) ∪ (right \ that) ∪ (that \ this)
     
     override def toString: String = s"[$left - [$elem] - $right]"    
     
-    override def equals(other: Any): Boolean = ???
+    override def equals(other: Any): Boolean =
```

Comment:
Boken contract on `hashCode`.

Reply:
Dear professor, I had add hashCode function. Please help me check it. Thanks.